### PR TITLE
Addition of capability to do force calculation using 3D CV method

### PR DESCRIPTION
**Pull request by gauravity30**
Hii,
I have implemented all the recommended changes in the previous PR. I have tried to keep the changes to minimum. I have also added a checking of errors after each mpi call. Instead of the creating new variables, I have used MPI_IN_PLACE. Kindly go through the changes and let me know if any further modification is required. I have tested the code for both Cylinder and Sphere at Re=3700 case. It is working fine. 

I have also created a new input file under the cylinder-wake case to represent the usage of new parameters (zfr, zbk, i2dsim, cez). By default the value of `i2dsim=1` i.e. the usual 2D CV method is used. If changed to 1, it used the new 3D CV method. 

Also I noticed an inconsitency in the formatting of the code, might be fixed in the later versions in some lines 2 spaces [(Example of 2 space)](https://github.com/xcompact3d/Incompact3d/blob/master/src/forces.f90#L28) indentation is being used whereas somewhere 3 spaces indentation is used [(Example of 3 space)](https://github.com/xcompact3d/Incompact3d/blob/master/src/forces.f90#L65).

Thanks, 
Gaurav

### CODE REVIEW

File: src/forces.f90 [hunk 1]
```diff
@@ -85,29 +89,35 @@ subroutine init_forces
...
        write(*,*) '========================Forces============================='
-       write(*,*) '                       (icvlf)      (icvrt) '
-       write(*,*) '                (jcvup) B____________C '
-       write(*,*) '                        \            \ '
-       write(*,*) '                        \     __     \ '
-       write(*,*) '                        \    \__\    \ '
-       write(*,*) '                        \            \ '
-       write(*,*) '                        \       CV   \ '
-       write(*,*) '                (jcvlw) A____________D '
+       write(*,*) '     (icvlf)      (icvrt)    (kcvbk)    (kcvfr)'
```
Comment (mathrack) [inline src/forces.f90]: Probably best to keep the old format when i2dsim = 1 and use the modified format only when i2dsim = 0 ?

File: src/forces.f90 [hunk 2]
```diff
@@ -256,267 +268,626 @@ subroutine force(ux1,uy1,ep1)
 
     call derx (ta1,ux1,di1,sx,ffx,fsx,fwx,xsize(1),xsize(2),xsize(3),0,ubcx)    ! dudx
     call derx (tb1,uy1,di1,sx,ffxp,fsxp,fwxp,xsize(1),xsize(2),xsize(3),1,ubcy) ! dvdx
+    call derx (tc1,uz1,di1,sx,ffxp,fsxp,fwxp,xsize(1),xsize(2),xsize(3),1,ubcz) ! dwdx
```
Comment (mathrack) [inline src/forces.f90]: Can we use te1 here ? It is best to have a PR that keep the 2D force calculation unchanged.

This operation is not needed when i2dsim = 1 (default). Move all the 3D-specific operations together and perform them only if needed.

File: src/forces.f90 [hunk 3]
```diff
@@ -256,267 +268,626 @@ subroutine force(ux1,uy1,ep1)
...
     call transpose_x_to_y(ta1,ta2) ! dudx
     call transpose_x_to_y(tb1,tb2) ! dvdx
 
     call transpose_x_to_y(ux1,ux2)
     call transpose_x_to_y(uy1,uy2)
+    call transpose_x_to_y(uz1, uz2)
```
Comment (mathrack) [inline src/forces.f90]: Idem

File: src/forces.f90 [hunk 4]
```diff
@@ -256,267 +268,626 @@ subroutine force(ux1,uy1,ep1)
...
     call transpose_x_to_y(ppi1,ppi2)
-
-    call dery (tc2,ux2,di2,sy,ffyp,fsyp,fwyp,ppy,ysize(1),ysize(2),ysize(3),1,ubcx) ! dudy
-    call dery (td2,uy2,di2,sy,ffy,fsy,fwy,ppy,ysize(1),ysize(2),ysize(3),0,ubcy)    ! dvdy
-    call transpose_y_to_x(tc2,tc1) ! dudy
-    call transpose_y_to_x(td2,td1) ! dvdy
+    call transpose_y_to_z(ux2, ux3)
+    call transpose_y_to_z(uy2, uy3)
+
+    call dery (td2,ux2,di2,sy,ffyp,fsyp,fwyp,ppy,ysize(1),ysize(2),ysize(3),1,ubcx) ! dudy
```
Comment (mathrack) [inline src/forces.f90:320]: Can we keep tc2 here ? It is best to have a PR that keep the 2D force calculation unchanged.

File: src/forces.f90 [hunk 5]
```diff
@@ -256,267 +268,626 @@ subroutine force(ux1,uy1,ep1)
...
+    call dery (te2,uy2,di2,sy,ffy,fsy,fwy,ppy,ysize(1),ysize(2),ysize(3),0,ubcy)    ! dvdy
```
Comment (mathrack) [inline src/forces.f90:321]: Can we keep td2 here ? It is best to have a PR that keep the 2D force calculation unchanged.

File: src/forces.f90 [hunk 6]
```diff
@@ -256,267 +268,626 @@ subroutine force(ux1,uy1,ep1)
...
+    call dery (tf2,uz2,di2,sy,ffyp,fsyp,fwyp,ppy,ysize(1),ysize(2),ysize(3),1,ubcz) ! dwdy
+    call transpose_y_to_x(td2,td1) ! dudy
```
Comment (mathrack) [inline src/forces.f90:322]: Store dudy in tc1. It is best to have a PR that keep the 2D force calculation unchanged.

File: src/forces.f90 [hunk 7]
```diff
@@ -256,267 +268,626 @@ subroutine force(ux1,uy1,ep1)
...
+    call transpose_y_to_x(te2,te1) ! dvdy
```
Comment (mathrack) [inline src/forces.f90:323]: Store dvdy in td1. It is best to have a PR that keep the 2D force calculation unchanged.

File: src/forces.f90 [hunk 8]
```diff
@@ -256,267 +268,626 @@ subroutine force(ux1,uy1,ep1)
...
+                  fac   = (onepfive*uy1(i,j,k)-two*uy01(i,j,k)+half*uy11(i,j,k))*(one-ep1(i,j,k))
+                  tsumy = tsumy+fac*dx*del_y(j+(xstart(2)-1))/dt !tsumy+fac*dx*dy/dt
+                  !sumy(k) = sumy(k)+dudt1*dx*dy
+               enddo
+            enddo
+            tunstxl(xstart(3)-1+k)=tsumx
+            tunstyl(xstart(3)-1+k)=tsumy
+         enddo
+         call MPI_ALLREDUCE(tunstxl,tunstx,nz,real_type,MPI_SUM,MPI_COMM_WORLD,code)
+         if (ierror /= 0) then
```
Comment (mathrack) [inline src/forces.f90]: The error code in the call above is inside the variable `code`. It is not relevant to check `ierror`.

This needs to be fixed at several locations

File: src/forces.f90 [hunk 9]
```diff
@@ -256,267 +268,626 @@ subroutine force(ux1,uy1,ep1)
...
+         !        endif
+         !     endif
+         if (nrank .eq. 0) then
+            write(38,*) t,xDrag_mean,yLift_mean
+            flush(38)
+         endif
+         if (mod(itime, icheckpoint).eq.0) then
+            if (nrank .eq. 0) then
+               write(filename,"('forces.dat',I7.7)") itime
+               call execute_command_line ("cp forces.dat "//filename)
```
Comment (mathrack) [inline src/forces.f90]: Is it possible to perform this `cp` operation only once after the `if i2dsim` / `elseif i2dsim` / `endif` ?

Currently this operation is performed multiple times when `nvol > 1`

File: src/forces.f90 [hunk 10]
```diff
@@ -256,267 +268,626 @@ subroutine force(ux1,uy1,ep1)
...
-             call execute_command_line ("cp forces.dat "//filename)
-          endif
-       endif
-    enddo
-
+    else
+       write(*,*) '==========================================================='
+       write(*,*) 'Error: Invalid value for the parameter i2dsim'
+       write(*,*) '==========================================================='
+       call MPI_ABORT(MPI_COMM_WORLD,code,ierror)
```
Comment (mathrack) [inline src/forces.f90]: Use `decomp_2d_abort` to abort the simulation.

Currently, in the PR, all ranks will print to stdout before calling `MPI_ABORT`, this is not a very good idea.

**Review by mathrack — CHANGES_REQUESTED** [review]
Hello,

Thank you for your PR. The formatting issue will be solved at the scale of the full code later using an automated script.

I have suggested some modifications. Please look at them.

Cheers

**Comment by mathrack** [comment]
The objective of the PR is to perform 2D force calculation on a 3D object ?

Do we recover the 2D force when `zfr(1) = -999999.0` and `zbk(1) = 999999.0` ?

**Comment by gauravity30** [comment]
> The objective of the PR is to perform 2D force calculation on a 3D object ?
> 
> Do we recover the 2D force when `zfr(1) = -999999.0` and `zbk(1) = 999999.0` ?

No, the objective is to perform force calculation for objects like Sphere or any other 3D body which is not span-wise symmetric like a Cylinder, or Airfoil.

**Comment by gauravity30** [comment]
> Hello,
> 
> Thank you for your PR. The formatting issue will be solved at the scale of the full code later using an automated script.
> 
> I have suggested some modifications. Please look at them.
> 
> Cheers

Cool, 
I had tried to keep the same format for derivatives as it is in the Case-Cylinder-Wake. Using different names for derivatives might not be a good idea. That's why I changed the names. 

I can shift the extra derivatives and operations inside the i3dsim==0 case as it is not required and affect the code performance unnecessarily. 

Thanks,
Gaurav

File: src/forces.f90 (hunk 1 above)
Comment (gauravity30) [reply src/forces.f90]: Got it. I will make the change.

File: src/forces.f90 (hunk 8 above)
Comment (gauravity30) [reply src/forces.f90]: I will fix it.

File: src/forces.f90 (hunk 9 above)
Comment (gauravity30) [reply src/forces.f90]: Will check if it is possible.

**Comment by mathrack** [comment]
> > The objective of the PR is to perform 2D force calculation on a 3D object ?
> > Do we recover the 2D force when `zfr(1) = -999999.0` and `zbk(1) = 999999.0` ?
> 
> No, the objective is to perform force calculation for objects like Sphere or any other 3D body which is not span-wise symmetric like a Cylinder, or Airfoil.

Please elaborate more clearly the objective of the PR. It looks like the output is a 2D force, whatever the value of `i2dsim`.

When the 3D force calculation is used on a 2D object with `zfr(1) = -999999.0` and `zbk(1) = 999999.0`, do we recover the 2D force calculation ?

**Comment by gauravity30** [comment]
> > > The objective of the PR is to perform 2D force calculation on a 3D object ?
> > > Do we recover the 2D force when `zfr(1) = -999999.0` and `zbk(1) = 999999.0` ?
> > 
> > No, the objective is to perform force calculation for objects like Sphere or any other 3D body which is not span-wise symmetric like a Cylinder, or Airfoil.
> 
> Please elaborate more clearly the objective of the PR. It looks like the output is a 2D force, whatever the value of `i2dsim`.
> 
> When the 3D force calculation is used on a 2D object with `zfr(1) = -999999.0` and `zbk(1) = 999999.0`, do we recover the 2D force calculation ?

The objective of the PR is to calculate Lift and Drag for 3D objects like a sphere or something which is not extruded in the the Z direction. 

The current 2d version only does calculation for objects extruded in Z direction. Refer to discussion #256 for more information. (https://github.com/xcompact3d/Incompact3d/discussions/256) 

If we set the zfr(1) = 0 and zbk(1)=Lz for a cylinder case, we will recover the same value of the drag ignoring effects due the vortex breakdown in the 3D case. If you run the calculation for a sphere, you will get different drag and lift values and the 3D one will give you the correct value. 

I hope it is clear.

**Comment by gauravity30** [comment]
Hello,
I have made the suggested changes in the PR.  
1. When `i2dsim==1`,  the extra derv and transpose functions are called.
2. When `i2dsim==1`,  modified format is printed in the log.
3. The checkpoint save operation is done outside the do loop as well as if else block once. The operation isn't required to be called for each CV as well doesn't depend on the type of method. 
4. Error code is checked after every MPI operation and `MPI_ABORT` was changed with `DECOMP_2D_ABORT`.

For the naming of the derivatives, I have kept the variable names same as that in the `Case-Cylinder-wake.f90`. It avoids confusion for further updates. 

Thanks,
Gaurav

**Comment by mathrack** [comment]
Thank you. It looks ok. I need a few days to check it.

**Comment by mathrack** [comment]
The current version of the PR is looking ok from my perspective. I think we could avoid duplicate and have some force calculation shared between the 2D and 3D parts. But this goes beyond the scope of the PR.

I have not yet tested the force calculation, I will try to do that during the next days. I have made some modifications in the PR. Thus it is probably best to wait for comments from another developer before merging.

**Comment by gauravity30** [comment]
> The current version of the PR is looking ok from my perspective. I think we could avoid duplicate and have some force calculation shared between the 2D and 3D parts. But this goes beyond the scope of the PR.
> 
> I have not yet tested the force calculation, I will try to do that during the next days. I have made some modifications in the PR. Thus it is probably best to wait for comments from another developer before merging.

Hii, 
The changes look good to me. I had checked the code on my local system as well as on HPC before your commits. It worked fine on both of them. Please check it once and let me know if any changes are required. 

I also wanted to discuss a few more stuff about xcompact in general (not related to the PR). 

Thanks,
Gaurav

**Review by rfj82982 — APPROVED** [review]
LGTM

**Pull request merged**